src/hooks/useConnectToLastUsedWallet.ts

```diff
@@ -1,38 +1,161 @@
-import { useState, useEffect } from 'react';
+import { useState, useEffect, useRef } from 'react';
 import type { Chain } from '@wormhole-foundation/sdk';
 import { TransferWallet } from 'utils/wallet';
 import useWalletProvider from 'hooks/useWalletProvider';
+import type { WalletData } from 'store/wallet';
 
-export const useConnectToLastUsedWallet = (
+function connectToChain(
+  chain: Chain | undefined,
+  wallet: TransferWallet,
+  setIsConnecting: (isConnecting: boolean) => void,
+  connectWallet: ReturnType<typeof useWalletProvider>['connectWallet'],
+  skipReconnect: boolean,
+) {
+  if (!chain || skipReconnect) {
+    setIsConnecting(false);
+    return;
+  }
+
+  let isUnmounted = false;
+
+  async function connect() {
+    setIsConnecting(true);
+
+    if (chain && !isUnmounted) {
+      try {
+        await connectWallet(chain, wallet, true);
+      } catch {
+        // Wallet connection errors are handled by the wallet provider
+        // We just need to ensure state is updated
+      }
+    }
+
+    if (!isUnmounted) {
+      setIsConnecting(false);
+    }
+  }
+
+  void connect();
+
+  return () => {
+    isUnmounted = true;
+  };
+}
+
+function useConnectToLastUsedWallet(
   sourceChain?: Chain,
   destChain?: Chain,
-): { isConnecting: boolean } => {
-  const { connectWallet } = useWalletProvider();
-  const [isConnecting, setIsConnecting] = useState(false);
+  sendingWallet?: WalletData,
+  receivingWallet?: WalletData,
+): { isConnecting: boolean } {
+  const { connectWallet, swapWallets } = useWalletProvider();
+  const [isConnectingSource, setIsConnectingSource] = useState(false);
+  const [isConnectingDestination, setIsConnectingDestination] = useState(false);
+
+  const prevSourceChainRef = useRef(sourceChain);
+  const prevDestChainRef = useRef(destChain);
+  const sendingWalletRef = useRef(sendingWallet);
+  const receivingWalletRef = useRef(receivingWallet);
+  const isSwappingRef = useRef(false);
+
+  // Update wallet refs on every render
+  sendingWalletRef.current = sendingWallet;
+  receivingWalletRef.current = receivingWallet;
+
+  // Check if both source and destination swapped
+  const sourceChanged = sourceChain !== prevSourceChainRef.current;
+  const destChanged = destChain !== prevDestChainRef.current;
+
+  const chainsSwapped =
+    sourceChanged &&
+    destChanged &&
+    sourceChain === prevDestChainRef.current &&
+    destChain === prevSourceChainRef.current;
 
```

Comment:
This file essentially has 5 side effects doing similar stuff 
1. when source chain changes (connect last source wallet)
3. when dest chain changes (connect last dest wallet)
4. when source wallet changes (connect dest wallet)
5. when destination wallet changes (connect source wallet)
6. when source and destination swap (just swap)